Hacking Strimzi for Cruise Control (#82)
strimzi/strimzi.github.io: _posts/2019-08-08-hacking-for-cruise-control.md

---
layout: post
title:  "Hacking Strimzi for Cruise Control"
date: 2019-08-08
author: kyle_liberti
---

Cruise Control is a higly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Controls’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working on a Stimzi deployed Kafka cluster.

# Prequisite Outline

Like any proof of concept, there are a few actions to consider when putting it together. For getting Cruise Control to work with Strimzi these actions are: 

- Aligning Kafka versions
- Gathering Kafka Metrics
- Accessing Zookeeper endpoint
- Containerizing and deploying Cruise Control

# Aligning Kafka versions

> Its worth noting that the latest Kafka version which Cruise Control supports at this time is 2.0.1. The latest version of Strimzi which supports Kafka version 2.0.1 is 0.11.4. Although its possible to run Cruise Control alongside with later versions of Kafka, 2.0.1+, it is not supported. Keep this in mind when deciding which version of Strimzi to deploy. This post will stick to using Strimzi 0.11.4, but feel free to experiment with more updated versions.  

Reviewer: Have you done any tests with Kafka 2.3? Pointing people to use fairly old version is not exactly helpful.

Author: I ran it without issue on Kafka 2.3 but I wasn't sure if I sure use Kafka 2.3 since the latest version that CC supports is 2.0.1. I could update the text and examples to use Kafka 2.3 if you think that makes more sense, I guess this is just a POC so it doesn't matter if CC official supports it or not

Author: Update to use Kafka 2.3

First create a Dockerfile like this:
```Dockerfile
FROM strimzi/kafka:0.11.4-kafka-2.0.1

ENV CC_VERSION=2.0.57
ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"

USER root

RUN yum -y install git java-1.8.0-openjdk-devel \
    && yum clean all -y

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar \
    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```

Then build the image:

```bash
docker build Dockerfile -t <kafka-image-name>
```

Create a basic Kafka resource, `kafka.yaml`, then specify the Cruise Control metrics reporter in the Kafka `metrics.reporters` config:
```yaml
apiVersion: kafka.strimzi.io/v0alpha1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    image: <kafka-image-name>

>     version: 1.0.1

Reviewer: ```suggestion
    version: 2.0.1
```

    replicas: 2
    listeners:
      plain: {}
      tls: {}
    config:

>       metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"

Reviewer: Can you add some more details how does this work? Does Prometheus need to be enabled? Or how will CC get the data from this?

Author: Yes, I'll flesh out the details in another commit

>       offsets.topic.replication.factor: 0

Reviewer: I'm quite sure that the `0` here is not standard. Does it even work with `0`?

Author: You are right, fixed it to one. Surprisingly it worked, it just defaulted to 1.

>       transaction.state.log.replication.factor: 0

Reviewer: Same as above.

      transaction.state.log.min.isr: 1

>       log.message.format.version: "1.0.1"

Reviewer: ```suggestion
      log.message.format.version: "2.0.1"
```

    storage:
      type: ephemeral
  zookeeper:

>     replicas: 2

Reviewer: `2` ZK replicas are very bad practice ... you should use 3.

---
layout: post
title:  "Hacking Strimzi for Cruise Control"
date: 2019-08-08
author: kyle_liberti
---

> Cruise Control is a higly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Controls’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.

Reviewer 2: ```suggestion
Cruise Control is a higly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Control’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.
```

> # Prequisite Outline

Reviewer 3: you have to use `<!--more-->` before this title in order to separate the first part that will be in the posts list as quick introduction.

Reviewer 2: ```suggestion
<!--more-->

# Prequisite Outline
```

Like any proof of concept, there are a few actions to consider when putting it together. For getting Cruise Control to work with Strimzi these actions are: 

- Aligning Kafka versions
- Gathering Kafka Metrics
- Accessing Zookeeper endpoint
- Containerizing and deploying Cruise Control

# Aligning Kafka versions

> Its worth noting that the latest Kafka version which Cruise Control supports at this time is 2.0.1. The latest version of Strimzi which supports Kafka version 2.0.1 is 0.11.4. Although its possible to run Cruise Control alongside with later versions of Kafka, 2.0.1+, it is not supported. Keep this in mind when deciding which version of Strimzi to deploy. This post will stick to using Strimzi 0.11.4, but feel free to experiment with more updated versions.

Reviewer 3: typo "Although it's...."

Reviewer 2: ```suggestion
It's worth noting that the latest Kafka version which Cruise Control officially supports at this time is 2.0.1. The latest version of Strimzi which supports Kafka version 2.0.1 is 0.11.4. Although it is possible to run Cruise Control with later versions of Kafka, 2.0.1+, it's not supported. Keep this in mind when deciding which version of Strimzi to deploy. This post will stick to using Strimzi 0.11.4, but feel free to experiment with more updated versions.
```

# Gathering Metrics

> In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka. Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components know as `metric reporters`. Kafka metric reporters designate what, where, when, and how Kafka metrics are exported. For instance, by default Kafka uses a JMX metric reporter to expose all of its metrics through JMX, by wrapping Kafka metric objects into MBeans. Upon creation of a new Kafka metric object, Kafka will put the metric object in a concurrent hash map for future tracking and pass the metric object to each running metric reporter for its own filtering, formating, and storage. Afterwards, the metric reporters will monitor the metric objects for Kafka stat updates.

Reviewer 2: Backticks (and these are backticks, right?) in markdown will be rendered as monospace/code. Probably safer to just use apostrophe.

```suggestion
In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka. Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components know as 'metric reporters'. Kafka metric reporters designate what, where, when, and how Kafka metrics are exported. For instance, by default Kafka uses a JMX metric reporter to expose all of its metrics through JMX, by wrapping Kafka metric objects into MBeans. Upon creation of a new Kafka metric object, Kafka will put the metric object in a concurrent hash map for future tracking and pass the metric object to each running metric reporter for its own filtering, formating, and storage. Afterwards, the metric reporters will monitor the metric objects for Kafka stat updates.
```

> Cruise Control ships its own custom metric reporter, the `Cruise Control metric reporter` but must be included in `/opt/kafka/libs/` directory of every Kafka broker in order for it to work. For making the Cruise Control metric reporter available to Strimzi Kafka brokers, one can create a new Kafka image from the following Dockerfile:

Reviewer 2: ```suggestion
Cruise Control ships its own custom metric reporter, the 'Cruise Control metric reporter' which must be included in the `/opt/kafka/libs/` directory of every Kafka broker in order for it to work. For making the Cruise Control metric reporter available to Strimzi Kafka brokers, you can create a new Kafka image from the following Dockerfile:
```

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar \
    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```

Build the image:

```bash
docker build . -t <kafka-image-name>
```

Then create a basic Kafka resource referencing the new Kafka image and specifying the `CruiseControlMetricsReporter` in the Kafka config one of the `metrics.reporters` using the following `kafka.yaml` file:
```yaml
apiVersion: kafka.strimzi.io/v0alpha1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    image: <kafka-image-name>
    version: 2.0.1
    replicas: 3
    listeners:
      plain: {}
      tls: {}
    config: 
      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      log.message.format.version: "2.0.1"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    userOperator: {}
    topicOperator: {}
```
Apply to cluster:

```bash
oc apply -f kafka.yaml
```

> After recieving metric objects directly from Kafka, the cruise control metric reporter will filter out undesired metric objects, and store the remaining in a concurrent hash map. Then at periodic intervals designated by the user, the reporter will loop through the concurrent hash map and wrap each metric object's value into a Cruise Control Metric object. The reporter will add information like the metric type, time of when the metric value was recorded, and the broker id of where the metric was taken from. Finally, then Cruise Control metric object is formatted into a byte array and stored back into Kafka within a topic designated for cruise control metics.

Reviewer 3: How does this metric reporter work? Is it really so different from the JMX exporter we are using in our Kafka images?
I read here that "...will filter out undesired metric objects..." so it seems to me that this CC metric reporter doesn't get just the metrics that CC needs but many more to not consider.

Reviewer 3: Typo "After receiving..."

Author: The Cruise Control metric reporter and other included metric reporters are executed as background threads on every broker at startup. The metric reporters track metric objects that are passed to them by the Kafka nodes and format the stats they receive from these metric objects into specific formats for export to other systems. For example, the Cruise Control metric reporter will transform the received metric objects into a Cruise Control metric objects and store the them in a Kafka topic for later Cruise Control processing. Since Cruise Control doesn't need all of the metric objects passed to it from Kafka for partition balancing, the Cruise Control metric reporter will only format and store the metric objects it needs from Kafka. The metric reporters _prepare_ the metrics for export.  For example, the JMX metric reporter wraps all Kafka metic objects into JMX MBeans so that java agents like the Prometheus JMX exporter(included in the Kafka images) can scrape the JMX MBeans for exporting to other systems like Prometheus.

> After recieving metric objects directly from Kafka, the cruise control metric reporter will filter out undesired metric objects, and store the remaining in a concurrent hash map. Then at periodic intervals designated by the user, the reporter will loop through the concurrent hash map and wrap each metric object's value into a Cruise Control Metric object. The reporter will add information like the metric type, time of when the metric value was recorded, and the broker id of where the metric was taken from. Finally, then Cruise Control metric object is formatted into a byte array and stored back into Kafka within a topic designated for cruise control metics.

Reviewer 2: ```suggestion
After receiving metric objects directly from Kafka, the cruise control metric reporter will filter out undesired metric objects, and store the remaining in a concurrent hash map. Then at periodic intervals designated by the user, the reporter will loop through the concurrent hash map and wrap each metric object's value into a Cruise Control Metric object. The reporter will add information like the metric type, time of when the metric value was recorded, and the broker id of where the metric was taken from. Finally, then Cruise Control metric object is formatted into a byte array and stored back into Kafka within a topic designated for Cruise Control metrics.
```

# Accessing Zookeeper

> Be default, communication between Zookeeper and all other Strimzi components must be encrytped. Using a hack by Jakub Scholz, one can create a Kubernetes service which reuses the Strimzi certs to pipe unecrypted Cruise Control traffic through to Zookeeper. This removes the need to include traffic encryting sidecars in the Cruise Control deployment.

Reviewer 3: Typo "By default..."

Reviewer 2: ```suggestion
By default, communication between Zookeeper and all other Strimzi components is encrypted. Using a hack by Jakub Scholz, you can create a Kubernetes service which reuses the Strimzi certs to pipe unencrypted Cruise Control traffic through to Zookeeper. This removes the need to include traffic encrypting sidecars in the Cruise Control deployment.
```

```
kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
```

# Containerizing and deploying Cruise Control

Create a Cruise Control image

```Dockerfile
FROM centos:7

ENV CC_VERSION=2.0.57
ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk

RUN yum -y install git java-1.8.0-openjdk-devel && \
    yum clean all -y

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

WORKDIR cruise-control

RUN ./gradlew jar copyDependantLibs

RUN sed -i 's/bootstrap.servers=.*/bootstrap.servers=my-cluster-kafka-bootstrap:9092/g' config/cruisecontrol.properties \
    && sed -i 's/zookeeper.connect=.*/zookeeper.connect=zoo-entrance:2181/g' config/cruisecontrol.properties \
    && sed -i 's/capacity.config.file=.*/capacity.config.file=config\/capacityJBOD.json/g' config/cruisecontrol.properties \

>     && sed -i 's/sample.store.topic.replication.factor=.*/sample.store.topic.replication.factor=1/g' config/cruisecontrol.properties 

Reviewer 3: Instead of "building" the configuration properties file here during the image build, is it possible to use a ConfigMap where to put the configuration and mounting that one as a volume in the CC deployment?

Author: Definitely possible, in fact, it would be the proper way to do this. For the blog post, I was hoping to keep the instructions down to a minimum, but I do see the utility in doing this (quick prototyping, testing, etc). You think its worth making this change?

RUN curl -L https://github.com/linkedin/cruise-control-ui/releases/download/v0.1.0/cruise-control-ui.tar.gz \
    -o /tmp/cruise-control-ui.tar.gz \
    && tar zxvf /tmp/cruise-control-ui.tar.gz

ENTRYPOINT ["/bin/bash", "-c", "./kafka-cruise-control-start.sh config/cruisecontrol.properties"]
```

```
docker build . -t <cruise-control-image-name>
```

Then create a Cruise Control deployment, `cruise-control-deployment.yaml`
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: my-cruise-control
  labels:
    app: cruise-control
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: cruise-control
        name: my-cruise-control
    spec:
      containers:
      - name: my-cruise-control
        image: <cruise-control-image-name>
        imagePullPolicy: 'Always'
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: cruise-control
    name: my-cruise-control
  name: my-cruise-control
spec:
  ports:
    - name: http-9090
      port: 9090
      protocol: TCP
      targetPort: 9090
  type: NodePort
  selector:
    name: my-cruise-control

```

```bash
oc apply -f cruise-control-deployment.yaml
```

# Interacting with Cruise Control API

> To access the Cruise Control REST API from outside out OKD, port-forward the Cruise Control service to localhost:9095:

Reviewer 3: It's better to talk about Kubernetes, not OKD.

Reviewer 2: ```suggestion
To access the Cruise Control REST API from outside Kubernetes, port-forward the Cruise Control service to localhost:9095:
```

```

> oc port-forward svc/my-cruise-control 9095:9090

Reviewer 3: Why not using 9090 even on localhost? Are you using in your test a port forwarding for Prometheus as well? I noticed that CC uses the same Prometheus port.

Author: I wanted to make a distinction between which port in the `oc` command was for the CC and which was for localhost. I wasn't using Prometheus so we could use 9090 for localhost here as well

---
layout: post
title:  "Hacking Strimzi for Cruise Control"
date: 2019-08-08
author: kyle_liberti
---

Cruise Control is a higly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Controls’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.

# Prequisite Outline

Like any proof of concept, there are a few actions to consider when putting it together. For getting Cruise Control to work with Strimzi these actions are: 

- Aligning Kafka versions
- Gathering Kafka Metrics
- Accessing Zookeeper endpoint
- Containerizing and deploying Cruise Control

# Aligning Kafka versions

> Its worth noting that the latest Kafka version which Cruise Control supports at this time is 2.0.1. The latest version of Strimzi which supports Kafka version 2.0.1 is 0.11.4. Although its possible to run Cruise Control alongside with later versions of Kafka, 2.0.1+, it is not supported. Keep this in mind when deciding which version of Strimzi to deploy. This post will stick to using Strimzi 0.11.4, but feel free to experiment with more updated versions.

Author: Reworded and updated to Kafka 2.3

# Gathering Metrics
In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka. Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components know as `metric reporters`. Kafka metric reporters designate what, where, when, and how metrics are exported. 

Kafka maintains two core types of metrics:
* **Yammer metrics** For _server-side_ metrics related to brokers. These metric objects are kept in a globally accessible Yammer metric registry which can be reached directly by metric reporters. 
* **Kafka metrics** For _client-side_ metrics related to Kafka producers, consumers, connect, and streams. These metric objects are tracked in concurrent hash map and explicitely passed to metric reporters when created. 

> The type of metrics exposed by the metric reporters, whether it be Yammer metrics, Kafka metrics, or both, is solely dependent on the implementation of the metric reporter. For example, the Kafka JMX metric reporter only wraps **Kafka metrics** as MBeans. This is counterintuitve since it's known that Kafka exposes all of its metrics, Kafka *and* Yammer, through JMX. The reason why the Kafka JMX metric reporter doesn't wrap **Yammer metrics** as MBeans is because they are already exposed as JMX MBeans through Kafka by default.

Author: Its a metric reporter that ships with Kafka and is enabled by default. I will make this more clear

Placing the Cruise Control metric reporter jar into the `/opt/kafka/libs/` directory of every Kafka broker allows Kafka to find the reporter at runtime. For the sake of simplicity, one can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:

```Dockerfile
FROM strimzi/kafka:0.11.4-kafka-2.0.1

ENV CC_VERSION=2.0.57
ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"

USER root

RUN yum -y install git java-1.8.0-openjdk-devel \
    && yum clean all -y

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar \
    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```
Then build the image:
```bash
docker build . -t <kafka-image-name>
```

> For the Cruise Control metric reporter to be activated at runtime, its package name must be listed in the `metrics.reporters` field of the Kafka config. For a Strimzi Kafka cluster, one can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' package name in the 'metrics.reporters' field using the following `kafka.yaml` file:

Author: I meant class name here

---
layout: post
title:  "Hacking Strimzi for Cruise Control"
date: 2019-08-08
author: kyle_liberti
---

> Cruise Control is a highly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Control’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.

Reviewer: _Kafka partition balancing_ sounds a bit strange to my ears. Could we maybe rephrase it to make it more clear what does it actually do?

> Cruise Control is a highly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Control’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.

Reviewer: _Cruise Control’s component pluggability makes it flexible to customize for different environments._ This sounds to me more like part of conclusion. At this point I do not know anything about any Cruise Control components. So maybe just _Cruise Control is flexible and easy to customize for different environments._?

(This is where sentence per line make life a looot easier ;-))

> Cruise Control is a highly customizable Kafka partition balancing and monitoring software run in production at Linkedin. Cruise Control’s component pluggability makes it flexible to customize for different environments. This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort. This post will walk through how to get Cruise Control working with a Stimzi deployed Kafka cluster.

Reviewer: Typo ... `Stimzi` => `Strimzi`.

Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime. For the sake of simplicity, one can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:

```Dockerfile
FROM strimzi/kafka:0.13.0-kafka-2.3.0

ENV CC_VERSION=2.0.57
ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"

USER root

RUN yum -y install git java-1.8.0-openjdk-devel \
    && yum clean all -y

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar \
    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```
Then building the image:
```bash

> docker build . -t <kafka-image-name>

Reviewer: This will nto work for most people with Minikube etc. We should at least mention that they will need to push it somewhere etc.

```yaml
apiVersion: kafka.strimzi.io/v0alpha1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    image: <kafka-image-name>
    version: 2.3.0
    replicas: 3
    listeners:
      plain: {}
      tls: {}
    config:
      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      log.message.format.version: "2.3"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    userOperator: {}
    topicOperator: {}
```
Then applying the Kafka custom resource to the cluster:

```bash
kubectl apply -f kafka.yaml
```

> The 'metrics.reporters' field can take a list of comma-separated metric reporter class names. At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.

Reviewer: I missing here the part how CC gets these metrics. Can we add something like _CC will read these metrics from the designated Kafka topic and use them for balancing. How to configure the connection between CC and Kafka will be descibed later in one of the next parts of this blog post._

Reviewer 2: At this point in the article the it's not yet been made clear the overall picture of what's going on. I think this whole metrics section could do with an outline at the start:

"""
Each broker is configured with a metric reporter. The metric reporter publishes a subset of the broker metrics to a special topic. Cruise Control consumes the metrics from this topic to learn about the load experienced by each broker in the cluster. 
"""

Then I think you can elaborate on each of those parts and the reader has a good enough mental model of the moving parts for it to all make sense.

# Deploying Cruise Control

With the Kafka cluster running and the Zookeeper endpoint accessible, Cruise Control can now be started safely and begin sampling metrics and monitoring the cluster. To run on Kubernetes, Cruise Control must be formatted into a container image. One can do this using the following Dockerfile. Note that the Cruise Control configuration file, 'cruisecontrol.properties', must be altered to reference the appropriate Kafka bootstrap and Zookeeper service endpoints. In addition to pointing Cruise Control to the proper service endpoints, the Dockerfile sets up the Cruise Control GUI frontend to be available upon startup.

```Dockerfile
FROM centos:7

ENV CC_VERSION=2.0.57
ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk

RUN yum -y install git java-1.8.0-openjdk-devel && \
    yum clean all -y

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

WORKDIR cruise-control

> RUN ./gradlew jar copyDependantLibs

Reviewer: Could we use a two stage build or something to make this part a bit less nasty? (such as having JDK and all build artifacts directly in the image)

Author: Updated to mult-stage builds

# Ensure Cruise Control log directory is writable
RUN mkdir logs \ 
    && touch ./logs/kafkacruisecontrol.log \
    && chmod a+rw -R .

# Install Cruise Control GUI Frontend
RUN curl -L https://github.com/linkedin/cruise-control-ui/releases/download/v0.1.0/cruise-control-ui.tar.gz \
    -o /tmp/cruise-control-ui.tar.gz \
    && tar zxvf /tmp/cruise-control-ui.tar.gz

ENTRYPOINT ["/bin/bash", "-c", "./kafka-cruise-control-start.sh config/cruisecontrol.properties"]
```
Build the image:
```
docker build . -t <cruise-control-image-name>
```
Then, using the following file, 'cruise-control-deployment.yaml', create a Cruise Control deployment that references the Cruise Control image, 'cruise-control-image-name', and a Cruise Control service endpoint for enabling communication with Cruise Control from outside the Cruise Control pod.
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: my-cruise-control
  labels:
    app: cruise-control
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: cruise-control
        name: my-cruise-control
    spec:
      containers:
      - name: my-cruise-control

>         image: <cruise-control-image-name>

Reviewer: This will nto work for most people with Minikube etc. We should at least mention that they will need to push it somewhere etc.

        imagePullPolicy: 'Always'
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: cruise-control
    name: my-cruise-control
  name: my-cruise-control
spec:
  ports:
    - name: http-9090
      port: 9090
      protocol: TCP
      targetPort: 9090

>   type: NodePort

Reviewer: You seem to use port-forward to access the service. So does this really need to be NodePort service?

---
layout: post
title:  "Hacking Strimzi for Cruise Control"
date: 2019-08-08
author: kyle_liberti
---

Cruise Control is a highly customizable software tool for simplifying the monitoring and balancing of workloads across a Kafka cluster.
The flexibility of Cruise Control makes it easy to customize for different environments and use cases.
This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort.
This post will walk through how to hook Cruise Control up to a Strimzi deployed Kafka cluster.
<!--more-->

# Prerequisite Outline

> Like any proof of concept, there are a few considerations and tradeoffs to think about before jumping in. For starters, one will need a Kafka cluster. At this time of this post, Cruise Control officially supports Kafka version 2.0.0 but should still be compatible with later versions of Kafka. This blog post will stick to using Kafka version 2.3.0 but feel free to experiment with other Kafka versions.

Reviewer 2: * If this blog is explaining to users how to they can set up CC for their cluster then describing it as proof of concept is going to put people off trying it.
* In a blog post it's better to talk about 'you', rather than 'one', since it's more understandable for people whose first language is not English.

```suggestion
There are a few considerations and tradeoffs to think about before jumping in. For starters, you will need a Kafka cluster. At the time of this post, Cruise Control officially supports Kafka version 2.0.0 but should still be compatible with later versions of Kafka. This blog post will stick to using Kafka version 2.3.0 but feel free to experiment with other Kafka versions.
```

> The core challenges of hacking Cruise Control to work with Strimzi are: 

Reviewer 2: ```suggestion
The core challenges of making Cruise Control work with Strimzi are: 
```

- Gathering Kafka Metrics for Cruise Control consumption
- Enabling communication between Cruise Control and Zookeeper
- Deploying Cruise Control in Kubernetes

> There are many ways of approaching these hurdles but this post will attempt to address these challenges with as few steps as possible.

Reviewer 2: I don't think we should shy away from saying something like
```suggestion
We're looking at having deeper Cruise Control integration in Strimzi in the future, but this post is about how to get Cruise Control working with Strimzi today. For this reason I will address these challenges in as few steps as possible.
```

# Gathering Metrics
In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka. Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components known as 'metric reporters'. Kafka metric reporters designate what, where, when, and how metrics are prepared for export. 

Kafka maintains two core types of metrics:
* **Yammer metrics** For _server-side_ metrics related to brokers. These metric objects are kept in a globally accessible Yammer metric registry which can be reached directly by metric reporters. 

> * **Kafka metrics** For _client-side_ metrics related to Kafka producers, consumers, connect, and streams. These metric objects are tracked in concurrent hash map and explicitly passed to metric reporters when created. 

Reviewer 2: Are these really the best terms? I know broker uses Yammer metrics, but conceptually they're just "broker metrics", and yammer is an implementation concern. That makes the other ones "client metrics".  And does CC even care about client metrics? If not then I don't think it's relevant to talk about client metrics at all, or just a sentence that CC doesn't care about them.

> The type of metrics exposed by the metric reporters, whether it be Yammer metrics, Kafka metrics, or both, is solely dependent on the implementation of the metric reporter. For example, the JMX metric reporter, a metric reporter provided and automatically enabled by Kafka, only wraps **Kafka metrics** as JMX MBeans. This is counterintuitive since it's known that Kafka exposes all of its metrics, Kafka *and* Yammer, through JMX. The reason why the Kafka JMX metric reporter doesn't wrap **Yammer metrics** as MBeans is because they are already exposed as JMX MBeans by Kafka by default.

Reviewer 2: I didn't understand this sentence after reading it, and I'm familiar with how metrics work in Kafka, so I think this needs work (assuming it's relevant enough to keep). But for me all you need is the next paragraph.

## Cruise Control Metric Reporter
Cruise Control ships its own implementation of a Kafka metric reporter, the 'Cruise Control metric reporter'. Unlike the JMX metric reporter, the Cruise Control metric reporter filters the metric objects it receives from Kafka, stowing the metric objects related to Kafka partition monitoring into a concurrent hash map. Then, at user-defined intervals, the reporter will loop through the concurrent hash map and create Cruise Control metric objects from the data provided by the metric objects. For the Cruise Control metric objects, the reporter will include information related to the metric type, value, time of recording, and broker origin. Finally, the reporter will format the Cruise Control metric object into a byte array and store it into a Kafka topic designated for Cruise Control metrics.

> Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime. For the sake of simplicity, one can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:

Reviewer 2: ```suggestion
Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime. You can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:
```

```Dockerfile
# Multi-stage build requires Docker 17.05 or higher
# --------------- Builder stage ---------------
FROM centos:7 AS builder

ENV CC_VERSION=2.0.57
ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"

RUN yum -y install git java-1.8.0-openjdk-devel

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar

# --------------- Final stage ---------------
FROM strimzi/kafka:0.13.0-kafka-2.3.0
COPY --from=builder cruise-control/cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```
Then building and pushing the image:
```bash
# When building, tag the image with the name of a container registry
# that is accessible by the Kubernetes cluster.
docker build . -t <registry>/<kafka-image-name>

# Push the image to that container registry
docker push <registry>/<kafka-image-name>
```

> For the Cruise Control metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config. For a Strimzi Kafka cluster, one can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' class name in the 'metrics.reporters' field using the following 'kafka.yaml' file:

Reviewer 2: ```suggestion
For the Cruise Control metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config. For a Strimzi Kafka cluster, you can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' class name in the 'metrics.reporters' field using the following 'kafka.yaml' file:
```

  name: my-cluster
spec:
  kafka:
    image: <registry>/<kafka-image-name>
    version: 2.3.0
    replicas: 3
    listeners:
      plain: {}
      tls: {}
    config:
      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      log.message.format.version: "2.3"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    userOperator: {}
    topicOperator: {}
```
Then applying the Kafka custom resource to the cluster:

```bash
kubectl apply -f kafka.yaml
```
The 'metrics.reporters' field can take a list of comma-separated metric reporter class names. At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.

At user-defined time windows, Cruise Control will read these reported metrics from the designated Kafka topic and use them for cluster monitoring and balancing. How to configure the connection between Cruise Control and Kafka will be described later in the post.

# Accessing Zookeeper

> Cruise Control relies on Zookeeper for the current state and configuration of the Kafka cluster. For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly. By default, communication between Zookeeper and all other Strimzi components must be encrypted and authenticated. However, using a hack by Jakub Scholz, one can get around this requirement by creating a Kubernetes service which reuses Strimzi certificates to pipe unencrypted and unauthenticated traffic to Zookeeper. This allows Cruise Control to communicate with Zookeeper without the need of proper authorization and encryption configuration via Java system properties or Stunnel sidecars.

Reviewer 2: ```suggestion
Cruise Control relies on Zookeeper for the current state and configuration of the Kafka cluster. For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly. By default, communication between Zookeeper and all other Strimzi components must be encrypted and authenticated. However, using a hack by Jakub Scholz, you can get around this requirement by creating a Kubernetes service which reuses Strimzi certificates to pipe unencrypted and unauthenticated traffic to Zookeeper. This allows Cruise Control to communicate with Zookeeper without the need of proper authorization and encryption configuration via Java system properties or Stunnel sidecars.
```

```
kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
```

**WARNING**: Allowing unauthenticated access to Zookeeper exposes sensitive Kafka metadata for reading and writing by any user. Nefarious users could exploit this access by altering topic configurations to mess up cluster state or by changing ACL access rules to elevate client privileges for stealing data.

# Deploying Cruise Control

> With the Kafka cluster running and the Zookeeper endpoint accessible, Cruise Control can now be started safely and begin sampling metrics and monitoring the cluster. To run on Kubernetes, Cruise Control must be formatted into a container image. One can do this using the following Dockerfile. Note that the Cruise Control configuration file, 'cruisecontrol.properties', must be altered to reference the appropriate Kafka bootstrap and Zookeeper service endpoints. In addition to pointing Cruise Control to the proper service endpoints, the Dockerfile sets up the Cruise Control GUI frontend to be available upon startup.

Reviewer 2: ```suggestion
With the Kafka cluster running and the Zookeeper endpoint accessible, Cruise Control can now be started safely and begin sampling metrics and monitoring the cluster. To run on Kubernetes, Cruise Control must be formatted into a container image. You can do this using the following Dockerfile. Note that the Cruise Control configuration file, 'cruisecontrol.properties', must be altered to reference the appropriate Kafka bootstrap and Zookeeper service endpoints. In addition to pointing Cruise Control to the proper service endpoints, the Dockerfile sets up the Cruise Control GUI frontend to be available upon startup.
```

# Push the image to that container registry
docker push <registry>/<cruise-control-image-name>
```
Then, using the following file, 'cruise-control-deployment.yaml', create a Cruise Control deployment that references the Cruise Control image, 'cruise-control-image-name', and a Cruise Control service endpoint for enabling communication with Cruise Control from outside the Cruise Control pod.
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: my-cruise-control
  labels:
    app: cruise-control
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: cruise-control
        name: my-cruise-control
    spec:
      containers:
      - name: my-cruise-control
        image: <registry>/<cruise-control-image-name>
        imagePullPolicy: 'Always'
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: cruise-control
    name: my-cruise-control
  name: my-cruise-control
spec:
  ports:
    - name: http-9090
      port: 9090
      protocol: TCP
      targetPort: 9090
  type: ClusterIP
  selector:
    name: my-cruise-control
```
Apply to the cluster:
```bash
kubectl apply -f cruise-control-deployment.yaml
```

# Interacting with Cruise Control API

> After Cruise Control is up and running, one can access the Cruise Control REST API from outside Kubernetes by port forwarding the Cruise Control service to 'localhost:9090'

Reviewer 2: ```suggestion
After Cruise Control is up and running, you can access the Cruise Control REST API from outside Kubernetes by port forwarding the Cruise Control service to 'localhost:9090'
```

```
kubectl port-forward svc/my-cruise-control 9090:9090
```

After forwarding the service, there are a few ways to interact with the Cruise Control API:

* curl
* cruise-control client
* Cruise Control GUI Frontend

## Using curl

```
curl -vv -X GET "http://127.0.0.1:9090/kafkacruisecontrol/state"
```

For more details visit the [Cruise Control Wiki](https://github.com/linkedin/cruise-control/wiki/REST-APIs)

## Using Cruise Control Python client

To install:

```
pip3 install cruise-control-client
```

To run:

```
cccli -a 127.0.0.1:9090  kafka_cluster_state
```

For more details visit the [Cruise Control Wiki](https://github.com/linkedin/cruise-control/wiki/Getting-Started)

## Using Cruise Control Frontend

Vist 'http://127.0.0.1:9090' in browser

![Cruise Control Frontend]({{ "/assets/2019-08-08-cruise-control-frontend.png" }})

# Conclusion

> After gathering enough metric samples, Cruise Control will be able estimate partition resources.  Knowing the the partition load, Cruise Control will be able to evaluate whether any resource restrictions have been violated and propose more efficient partition assignments for the cluster. Just like metric reporters, many of the Cruise Control's components can be tailored to fit special use cases and environments. Components that determine things like how resources are calculated, what resource goals are prioritized, and how anomalies are handled can all be customized by the user. This flexibility gives users the leverage to operate their Kafka clusters more effectively, even while using Strimzi. In one of the future versions, the Strimzi community hopes to make this even easier by supporting Cruise Control directly from the Strimzi Operator, but until then, hack away!

Reviewer 2: ```suggestion
After gathering enough metric samples, Cruise Control will be able estimate partition resources.  Knowing the partition load, Cruise Control will be able to evaluate whether any resource restrictions have been violated and propose more efficient partition assignments for the cluster. Just like metric reporters, many of Cruise Control's components can be tailored to fit special use cases and environments. Components that determine things like how resources are calculated, what resource goals are prioritized, and how anomalies are handled can all be customized by the user. This flexibility gives users the leverage to operate their Kafka clusters more effectively, even while using Strimzi. In one of the future versions, the Strimzi community hopes to make this even easier by supporting Cruise Control directly from the Strimzi Operator, but until then, hack away!
```

This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort.
This post will walk through how to hook Cruise Control up to a Strimzi deployed Kafka cluster.
<!--more-->

# Prerequisite Outline

There are a few considerations and tradeoffs to think about before jumping in.
For starters, you will need a Kafka cluster.
At the time of this post, Cruise Control officially supports Kafka version 2.0.0 but should still be compatible with later versions of Kafka.
This blog post will stick to using Kafka version 2.3.0 but feel free to experiment with other Kafka versions.

The core challenges of making Cruise Control work with Strimzi are: 

- Gathering Kafka Metrics for Cruise Control consumption
- Enabling communication between Cruise Control and Zookeeper
- Deploying Cruise Control in Kubernetes

We're looking at having deeper Cruise Control integration in Strimzi in the future, but this post is about how to get Cruise Control working with Strimzi today. For this reason I will address these challenges in as few steps as possible.

> # Gathering Metrics

Reviewer 3: blank line after the title

In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka.
Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components known as 'metric reporters'. 
Kafka metric reporters designate what, where, when, and how metrics are prepared for export.
For Cruise Control, each Kafka broker is equipped with a metric reporter that will publish a subset of the broker's metrics to a special topic. 
Cruise Control will then consume the metrics from this topic to learn about the load experienced by each broker in the cluster.

Upon startup, every Kafka server loops through a list of metric reporters supplied via its Kafka config, executing every metric reporter on its own thread.
As Kafka creates metric objects it passes them to every metric reporter thread for filtering and tracking.
What, where, and how this happens to the metric objects depends on the specific implementation of the metric reporter.
Afterwards, the metric reporters monitor the metric objects for stat updates from Kafka, formatting and storing the changes for future export.

> ## Cruise Control Metric Reporter

Reviewer 3: blank line after the title

Cruise Control ships its own implementation of a Kafka metric reporter, the 'Cruise Control metric reporter'.

> The Cruise Control metric reporter filters the metric objects it receives from Kafka, stowing the metric objects related to Kafka partition monitoring into a concurrent hash map.

Reviewer 3: typo "stowing"

      log.message.format.version: "2.3"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    userOperator: {}
    topicOperator: {}
```
Then applying the Kafka custom resource to the cluster:

```bash
kubectl apply -f kafka.yaml
```
The `metrics.reporters` field can take a list of comma-separated metric reporter class names, if you need to add others.
At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.

At user-defined time windows, Cruise Control will read these reported metrics from the designated Kafka topic and use them for cluster monitoring and balancing.
How to configure the connection between Cruise Control and Kafka will be described later in the post.

# Accessing Zookeeper

Cruise Control gets the current state and configuration of the Kafka cluster from Zookeeper.
For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly.
Communication between Zookeeper and all other Strimzi components is always encrypted and authenticated using TLS.

> This preventsCruise Control making a non-TLS connection  to Zookeeper.

Reviewer 3: adding a space "This prevents Cruise Control...."

Communication between Zookeeper and all other Strimzi components is always encrypted and authenticated using TLS.
This preventsCruise Control making a non-TLS connection  to Zookeeper.
However, using a hack by Jakub Scholz, you can get around this requirement by creating a Kubernetes service which reuses Strimzi certificates to pipe unencrypted and unauthenticated traffic to Zookeeper. This allows Cruise Control to communicate with Zookeeper without the need of proper authorization and encryption configuration via Java system properties or Stunnel sidecars.

```
kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
```

**WARNING**: Allowing unauthenticated access to Zookeeper exposes sensitive Kafka metadata for reading and writing by any user.
Nefarious users could exploit this access by altering topic configurations to mess up cluster state or by changing ACL access rules to elevate client privileges for stealing data.

**Note**: Kafka is in the process of migrating to using the Kafka `AdminClient` instead of Zookeeper for operations like initiating replica reassignment.

> To stay compatible with future versions of Kafka, Cruise Control will need to follow these API changes.

Author: @tombentley will these two lines suffice for mention of the AdminClient?

Reviewer 2: TBH I think just the first is sufficient.

The core challenges of making Cruise Control work with Strimzi are: 

- Gathering Kafka Metrics for Cruise Control consumption
- Enabling communication between Cruise Control and Zookeeper
- Deploying Cruise Control in Kubernetes

We're looking at having deeper Cruise Control integration in Strimzi in the future, but this post is about how to get Cruise Control working with Strimzi today. For this reason I will address these challenges in as few steps as possible.

# Gathering Metrics
In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka.
Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components known as 'metric reporters'. 
Kafka metric reporters designate what, where, when, and how metrics are prepared for export.
For Cruise Control, each Kafka broker is equipped with a metric reporter that will publish a subset of the broker's metrics to a special topic. 
Cruise Control will then consume the metrics from this topic to learn about the load experienced by each broker in the cluster.

> Kafka maintains two core types of metrics:

Reviewer 2: Although this is a good description of metric reporters themselves, CC is all about the broker, and so you only need to explain enough about metric reporter _as used by the broker_ for the reader to understand the context behind why they're configuring this thing. I think you nailed that in the previous paragraph and this and the following paragraphs in this section really don't add much that the reader needs to know. (I can see that your enthusiastic to share what you've learned about metric reports, which is fine, but I don't think this is the blog post to do that – if you digress too much the reader might stop reading).

"""

Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime.
You can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:

```Dockerfile
# Multi-stage build requires Docker 17.05 or higher
# --------------- Builder stage ---------------
FROM centos:7 AS builder

ENV CC_VERSION=2.0.57
ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"

RUN yum -y install git java-1.8.0-openjdk-devel

RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git

RUN cd cruise-control \
    && ./gradlew jar :cruise-control-metric-reporter:jar

# --------------- Final stage ---------------
FROM strimzi/kafka:0.13.0-kafka-2.3.0
COPY --from=builder cruise-control/cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
```
Then building and pushing the image:
```bash
# When building, tag the image with the name of a container registry
# that is accessible by the Kubernetes cluster.
docker build . -t <registry>/<kafka-image-name>

# Push the image to that container registry
docker push <registry>/<kafka-image-name>
```

> For the Cruise Control metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config.

Reviewer 2: ```suggestion
For a metric reporter to be activated at runtime, its class name must be listed in the 'metrics.reporters' field of the Kafka config.
```

> For a Strimzi Kafka cluster, you can do this by creating a Kafka custom resource which references the new Kafka 'image' built above and the 'com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter' class name in the 'metrics.reporters' field using the following 'kafka.yaml' file:

Reviewer 2: ```suggestion
You do this in Strimzi by creating a `Kafka` custom resource which references the new Kafka 'image' built above and the `com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter` class name in the `metrics.reporters` field using the following 'kafka.yaml' file:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    image: <registry>/<kafka-image-name>
    version: 2.3.0
    replicas: 3
    listeners:
      plain: {}
      tls: {}
    config:
      metric.reporters: "com.linkedin.kafka.cruisecontrol.metricsreporter.CruiseControlMetricsReporter"
      offsets.topic.replication.factor: 1
      transaction.state.log.replication.factor: 1
      transaction.state.log.min.isr: 1
      log.message.format.version: "2.3"
    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    userOperator: {}
    topicOperator: {}
```
Then applying the Kafka custom resource to the cluster:

```bash
kubectl apply -f kafka.yaml
```

> The 'metrics.reporters' field can take a list of comma-separated metric reporter class names.

Reviewer 2: ```suggestion
The `metrics.reporters` field can take a list of comma-separated metric reporter class names, if you need to add others.
```

At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.

At user-defined time windows, Cruise Control will read these reported metrics from the designated Kafka topic and use them for cluster monitoring and balancing.
How to configure the connection between Cruise Control and Kafka will be described later in the post.

# Accessing Zookeeper

> Cruise Control relies on Zookeeper for the current state and configuration of the Kafka cluster.

Reviewer 2: ```suggestion
Cruise Control gets the current state and configuration of the Kafka cluster from Zookeeper.
```

I think we should mention that they are in the process of migrating to using the Kafka `AdminClient`.

For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly.

> By default, communication between Zookeeper and all other Strimzi components must be encrypted and authenticated.

Reviewer 2: ```suggestion
Communication between Zookeeper and all other Strimzi components is always encrypted and authenticated using TLS.
This preventsCruise Control making a non-TLS connection  to Zookeeper.
```
